internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -100,159 +111,217 @@ func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineD
-
-	minReadySeconds := int32(0)
-	if d.Spec.MinReadySeconds != nil {
-		minReadySeconds = *d.Spec.MinReadySeconds
-	}
-
-	// Create new MachineSet
-	newMS := clusterv1.MachineSet{
-		ObjectMeta: metav1.ObjectMeta{
-			// Make the name deterministic, to ensure idempotence
-			Name:      d.Name + "-" + apirand.SafeEncodeString(machineTemplateSpecHash),
-			Namespace: d.Namespace,
-			Labels:    make(map[string]string),
-			// Note: by setting the ownerRef on creation we signal to the MachineSet controller that this is not a stand-alone MachineSet.
-			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(d, machineDeploymentKind)},
-		},
-		Spec: clusterv1.MachineSetSpec{
-			ClusterName:     d.Spec.ClusterName,
-			Replicas:        new(int32),
-			MinReadySeconds: minReadySeconds,
-			Selector:        *newMSSelector,
-			Template:        newMSTemplate,
-		},
-	}
-
-	// Set the labels from newMSTemplate as top-level labels for the new MS.
-	// Note: We can't just set `newMSTemplate.Labels` directly and thus "share" the labels map between top-level and
-	// .spec.template.metadata.labels. This would mean that adding the MachineDeploymentNameLabel later top-level
-	// would also add the label to .spec.template.metadata.labels.
-	for k, v := range newMSTemplate.Labels {
-		newMS.Labels[k] = v
+	patchOptions := []client.PatchOption{
+		client.ForceOwnership,
+		client.FieldOwner(machineDeploymentManagerName),
 	}
-
-	// Enforce that the MachineDeploymentNameLabel label is set
-	// Note: the MachineDeploymentNameLabel is added by the default webhook to MachineDeployment.spec.template.labels if spec.selector is empty.
-	newMS.Labels[clusterv1.MachineDeploymentNameLabel] = d.Name
-
-	if d.Spec.Strategy.RollingUpdate.DeletePolicy != nil {
-		newMS.Spec.DeletePolicy = *d.Spec.Strategy.RollingUpdate.DeletePolicy
+	err = r.Client.Patch(ctx, newMS, client.Apply, patchOptions...)
+	if err != nil {
+		log.Error(err, "Failed to create new MachineSet", "MachineSet", klog.KObj(newMS))
+		r.recorder.Eventf(d, corev1.EventTypeWarning, "FailedCreate", "Failed to create MachineSet %q: %v", newMS.Name, err)
+		return nil, err
 	}
+	log.V(4).Info("Created new MachineSet", "MachineSet", klog.KObj(newMS))
+	r.recorder.Eventf(d, corev1.EventTypeNormal, "SuccessfulCreate", "Created MachineSet %q", newMS.Name)
 
-	// Add foregroundDeletion finalizer to MachineSet if the MachineDeployment has it
-	finalizerSet := sets.Set[string]{}.Insert(d.Finalizers...)
-	if finalizerSet.Has(metav1.FinalizerDeleteDependents) {
-		controllerutil.AddFinalizer(&newMS, metav1.FinalizerDeleteDependents)
+	// Keep trying to get the MS. This will force the cache to update and prevent any future reconciliation of the
+	// MachineDeployment to reconcile with an outdated list of MachineSets which could lead to unwanted creation of a
+	// duplicate MachineSet.
+	if err := wait.PollImmediate(stateConfirmationInterval, stateConfirmationTimeout, func() (bool, error) {
```

Comment:
nit: I think the constant names are a bit confusing. I think we don't need them as constants to be honest. They are only used here and nobody else cares about them